Project 1

- In this project I am using a csv file of sephora products that I found online
- I modify the file and write it to an SQL database
- I also write the modified file to a new csv file
- After converting this file to an SQL database I give some summaries of the file

In [111]:
import sys
!{sys.executable} -m pip install PyMySQL

import sys
!{sys.executable} -m pip install mysql.connector

In [112]:
import pandas as pd
import os
import pymysql
import mysql.connector
import matplotlib.pyplot as plt
from sqlalchemy import create_engine

Connect to SQL server

In [113]:
host_name = "localhost"
host_ip = "127.0.0.1"
port = "3306"

user_id = "root"
pwd = "Kcdswims12"

src_dbname = "northwind"
dst_dbname = "northwind_dw2"

Read in the file

In [159]:
try:
    mydata = pd.read_csv("sephora_website_dataset.csv")
except:
    print('Error occurred. Make sure file is in correct directory')

mydata.head(2)


id           brand   category                            name  \
0  2218774  Acqua Di Parma  Fragrance  Blu Mediterraneo MINIATURE Set   
1  2044816  Acqua Di Parma    Cologne                         Colonia   

             size  rating  number_of_reviews  love  price  value_price  ...  \
0  5 x 0.16oz/5mL     4.0                  4  3002   66.0         75.0  ...   
1   0.7 oz/ 20 mL     4.5                 76  2700   66.0         66.0  ...   

  MarketingFlags  MarketingFlags_content  \
0           True             online only   
1           True             online only   

                                             options  \
0                                         no options   
1   - 0.7 oz/ 20 mL  Spray  - 1.7 oz/ 50 mL Eau d...   

                                             details  \
0  This enchanting set comes in a specially handc...   
1  An elegant timeless scent filled with a fresh-...   

                                          how_to_use  \
0  Suggested Usage:-Fragrance is intensified by t...   
1                                    no instructions   

                                         ingredients online_only  exclusive  \
0  Arancia di Capri Eau de Toilette: Alcohol Dena...           1          0   
1                                            unknown           1          0   

   limited_edition  limited_time_offer  
0                0                   0  
1                0                   0  

[2 rows x 21 columns]

Convert to dataframe

In [115]:
df = pd.DataFrame(mydata)
df.head(2)

id           brand   category                            name  \
0  2218774  Acqua Di Parma  Fragrance  Blu Mediterraneo MINIATURE Set   
1  2044816  Acqua Di Parma    Cologne                         Colonia   

             size  rating  number_of_reviews  love  price  value_price  ...  \
0  5 x 0.16oz/5mL     4.0                  4  3002   66.0         75.0  ...   
1   0.7 oz/ 20 mL     4.5                 76  2700   66.0         66.0  ...   

  MarketingFlags  MarketingFlags_content  \
0           True             online only   
1           True             online only   

                                             options  \
0                                         no options   
1   - 0.7 oz/ 20 mL  Spray  - 1.7 oz/ 50 mL Eau d...   

                                             details  \
0  This enchanting set comes in a specially handc...   
1  An elegant timeless scent filled with a fresh-...   

                                          how_to_use  \
0  Suggested Usage:-Fragrance is intensified by t...   
1                                    no instructions   

                                         ingredients online_only  exclusive  \
0  Arancia di Capri Eau de Toilette: Alcohol Dena...           1          0   
1                                            unknown           1          0   

   limited_edition  limited_time_offer  
0                0                   0  
1                0                   0  

[2 rows x 21 columns]

In [146]:
df = df.drop_duplicates(subset='id', keep='first', inplace=False)

Create a new database in mysql

In [160]:
try:
    mydb = pymysql.connect(
      host=host_name,
      user=user_id,
      password=pwd
    )
    mycursor = mydb.cursor()
    mycursor.execute("DROP DATABASE IF EXISTS mydatabase")
    mycursor.execute("CREATE DATABASE mydatabase")
except:
    print('Error occurred. Make sure local instance is on.')

In [161]:
mydb = pymysql.connect(
  host=host_name,
  user=user_id,
  password=pwd,
  database = "mydatabase"
)
mycursor = mydb.cursor()

Write the dataframe to the new database in a table labelled 'products'

In [162]:
engine = create_engine("mysql+pymysql://{user}:{pw}@localhost/{db}"
                       .format(user=user_id,
                               pw=pwd,
                               db="mydatabase"))

In [163]:
df.to_sql('products', con = engine, if_exists = 'append', chunksize = 1000)

Modifying the columns (adding/removing, renaming)

In [164]:
def get_dataframe(user_id, pwd, host_name, mydatabase, sql_query):
    conn_str = f"mysql+pymysql://{user_id}:{pwd}@{host_name}/{mydatabase}"
    sqlEngine = create_engine(conn_str, pool_recycle=3600)
    connection = sqlEngine.connect()
    dframe = pd.read_sql(sql_query, connection);
    connection.close()
    
    return dframe

In [165]:
def set_dataframe(user_id, pwd, host_name, db_name, df, table_name, pk_column, db_operation):
    conn_str = f"mysql+pymysql://{user_id}:{pwd}@{host_name}/{db_name}"
    sqlEngine = create_engine(conn_str, pool_recycle=3600)
    connection = sqlEngine.connect()
    
    if db_operation == "insert":
        df.to_sql(table_name, con=connection, index=False, if_exists='replace')
        sqlEngine.execute(f"ALTER TABLE {table_name} ADD PRIMARY KEY ({pk_column});")
            
    elif db_operation == "update":
        df.to_sql(table_name, con=connection, index=False, if_exists='append')
    
    connection.close()

In [166]:
sql_products = "SELECT * FROM mydatabase.products;"
df_products = get_dataframe(user_id, pwd, host_name, "mydatabase", sql_products)

In [167]:
colnames = df.columns.values.tolist()
colnames

['id',
 'brand',
 'category',
 'name',
 'size',
 'rating',
 'number_of_reviews',
 'love',
 'price',
 'value_price',
 'URL',
 'MarketingFlags',
 'MarketingFlags_content',
 'options',
 'details',
 'how_to_use',
 'ingredients',
 'online_only',
 'exclusive',
 'limited_edition',
 'limited_time_offer']

In [168]:
drop_cols = ['number_of_reviews',
             'love',
             'value_price',
             'URL',
             'MarketingFlags',
             'MarketingFlags_content',
             'options',
             'details',
             'how_to_use',
             'ingredients',
             'limited_time_offer'
            ]
df_products.drop(drop_cols, axis=1, inplace=True)
#df_products.rename(columns={"id":"employee_key"}, inplace=True)
df_products.head(2)

index       id           brand   category                            name  \
0      0  2218774  Acqua Di Parma  Fragrance  Blu Mediterraneo MINIATURE Set   
1      1  2044816  Acqua Di Parma    Cologne                         Colonia   

             size  rating  price  online_only  exclusive  limited_edition  
0  5 x 0.16oz/5mL     4.0   66.0            1          0                0  
1   0.7 oz/ 20 mL     4.5   66.0            1          0                0

In [169]:
primary_key="id"
db_operation="insert"
set_dataframe(user_id, pwd, host_name, 'mydatabase', df_products, 'products', primary_key, db_operation)

In [174]:
df = pd.read_sql("SELECT * FROM products", mydb)
mydb.close()

Number of records

In [175]:
len(df)

9115

Number of columns

In [176]:
len(df.columns)

11

Statistical Summaries

In [177]:
numerical_cols = [col for col in df.columns if df.dtypes[col] != 'O']
categorical_cols = [col for col in df.columns if col not in numerical_cols]

print(numerical_cols)
print(categorical_cols)

['index', 'id', 'rating', 'price', 'online_only', 'exclusive', 'limited_edition']
['brand', 'category', 'name', 'size']


In [178]:
df[numerical_cols].describe()

index            id       rating        price  online_only  \
count  9115.000000  9.115000e+03  9115.000000  9115.000000  9115.000000   
mean   4581.179046  1.963309e+06     3.992101    50.214016     0.235217   
std    2645.874874  3.859365e+05     1.004239    47.231030     0.424157   
min       0.000000  5.000000e+01     0.000000     2.000000     0.000000   
25%    2291.500000  1.819800e+06     4.000000    24.000000     0.000000   
50%    4581.000000  2.072973e+06     4.000000    35.000000     0.000000   
75%    6868.500000  2.230803e+06     4.500000    59.000000     0.000000   
max    9167.000000  2.359685e+06     5.000000   549.000000     1.000000   

         exclusive  limited_edition  
count  9115.000000      9115.000000  
mean      0.264070         0.092046  
std       0.440861         0.289107  
min       0.000000         0.000000  
25%       0.000000         0.000000  
50%       0.000000         0.000000  
75%       1.000000         0.000000  
max       1.000000         1.000000

Observations:

- The average price of products is $50.214016

- The average rating of products is 3.992101

- About 26% of products are exclusive

- About 9% of products are limited edition

- About 23.5% of products are online only

Writing the new dataframe out to a csv file

In [157]:
data_dir = os.path.join(os.getcwd(), 'data')
dest_file = os.path.join(data_dir, 'project1.csv')

df.to_csv(dest_file)

SCRAP WORK

mycursor.execute('''
        CREATE TABLE products (
            id int primary key,
            brand varchar(50),
            category varchar(50),
            name varchar(50),
            size varchar(50),
            rating int,
            price int,
            online_only int,
            exclusive int,
            limited_edition int
            )
               ''')
               
for row in df():
    mycursor.execute('''
                INSERT INTO products (
                id, 
                brand,
                category,
                name,
                size,
                rating,
                price,
                online_only,
                exclusive,
                limited_edition
                )
                VALUES (?,?,?,?,?,?,?,?,?,?)
                ''',
                row.id, 
                row.brand,
                row.category,
                row.name,
                row.size,
                row.rating,
                row.price,
                row.online_only,
                row.exclusive,
                row.limited_edition
                )
mydb.commit()




'id',
 'brand',
 'category',
 'name',
 'size',
 'rating',
 'price',
 'online_only',
 'exclusive',
 'limited_edition',
 
 
 primary_key="id"
db_operation="insert"
set_dataframe(user_id, pwd, host_name, 'mydatabase', df_products, 'products', primary_key, db_operation)

In [ ]:
mydb.close()